lake/branch.go

```diff
@@ -0,0 +1,441 @@
+package lake
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/brimdata/zed/expr/extent"
+	"github.com/brimdata/zed/lake/commit"
+	"github.com/brimdata/zed/lake/index"
+	"github.com/brimdata/zed/lake/journal"
+	"github.com/brimdata/zed/lake/segment"
+	"github.com/brimdata/zed/order"
+	"github.com/brimdata/zed/pkg/nano"
+	"github.com/brimdata/zed/pkg/storage"
+	"github.com/brimdata/zed/zio"
+	"github.com/brimdata/zed/zio/zngio"
+	"github.com/brimdata/zed/zng"
+	"github.com/brimdata/zed/zson"
+	"github.com/segmentio/ksuid"
+)
+
+type BranchConfig struct {
+	Name   string      `zng:"name"`
+	ID     ksuid.KSUID `zng:"id"`
+	Parent ksuid.KSUID `zng:"parent"`
+}
+
+type Branch struct {
+	BranchConfig
+	pool   *Pool
+	engine storage.Engine
+	log    *commit.Log
+	base   commit.View
+}
+
+func newBranchConfig(name string, parent ksuid.KSUID) *BranchConfig {
+	//XXX need to take parent args and check for a cycle.
+	// also check that parent journal ID exists in parent.
+	return &BranchConfig{
+		Name:   name,
+		ID:     ksuid.New(),
+		Parent: parent,
+	}
+}
+
+func (b *BranchConfig) Path(poolPath *storage.URI) *storage.URI {
+	return poolPath.AppendPath(b.ID.String())
+}
+
+func (b *BranchConfig) Create(ctx context.Context, engine storage.Engine, poolPath *storage.URI, order order.Which, base journal.ID) error {
+	_, err := commit.Create(ctx, engine, b.Path(poolPath), order, base)
+	return err
+}
+
+func (b *BranchConfig) Open(ctx context.Context, engine storage.Engine, parent *storage.URI, pool *Pool) (*Branch, error) {
```

Comment:
I think poolPath makes this clearer, and is consistent with the Create call:
```suggestion
func (b *BranchConfig) Open(ctx context.Context, engine storage.Engine, poolPath *storage.URI, pool *Pool) (*Branch, error) {
```